File: BPMOPMobile_VietNamAirLines/BPMOPMobile_VietNamAirLines/BPMOPMobile.Droid.Core/Component/ControlDate.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;
using System.Text;

using Android.App;
using Android.Content;
using Android.Graphics;
using Android.OS;
using Android.Runtime;
using Android.Support.V4.Content;
using Android.Util;
using Android.Views;
using Android.Widget;
using BPMOPMobile.Bean;
using BPMOPMobile.Class;
using BPMOPMobile.Droid.Core.Class;
using BPMOPMobile.Droid.Core.Common;

namespace BPMOPMobile.Droid.Core.Component
{
    public class ControlDate : ControlBase
    {
        public LinearLayout _parentView { get; set; }
        private ViewElement _element { get; set; }
        public override string Value
        {
            set
            {
                if (!String.IsNullOrEmpty(value))
                {
                    var _customValue = "";
                    DateTime dateValue = new DateTime();

                    try
                    {
                        //dateValue = DateTime.Parse(value, new CultureInfo("en", false));
                        dateValue = DateTime.Parse(value);
                    }
                    catch (FormatException)
                    {
                        Console.WriteLine("Unable to convert '{0}'.", value);
                    }

                    if (_element.DataType == "date")
                    {
                        if (BPMOPMobile.Class.CmmVariable.SysConfig.LangCode.Equals(CmmDroidVariable.M_SysLangVN))
                            _customValue = dateValue.ToString(@"dd/MM/yy", new CultureInfo("en-US"));
                        else
                            _customValue = dateValue.ToString(@"MM/dd/yy", new CultureInfo("en-US"));
                    }
                    else if (_element.DataType == "datetime")
                    {
                        ////_customValue = dateValue.ToString(@"dd/MM/yyyy hh:mm tt", new CultureInfo("en-US")).ToLower();
                        if (BPMOPMobile.Class.CmmVariable.SysConfig.LangCode.Equals(CmmDroidVariable.M_SysLangVN))
                            _customValue = dateValue.ToString(@"dd/MM/yy HH:mm").ToLower();
                        else
                            _customValue = dateValue.ToString(@"MM/dd/yy HH:mm").ToLower();
                    }
                    else if (_element.DataType == "time")
                    {
                        ////_customValue = dateValue.ToString("HH:mm", new CultureInfo("en-US"));
                        _customValue = dateValue.ToString("HH:mm");
                    }
                    this._tvValue.Text = _customValue;
                }
                else
                {
                    if (_element.Enable)
                    {
                        if (_element.DataType == "date")
                        {
                            this._tvValue.Text = CmmFunction.GetTitle("TEXT_CHOOSE_DATE", "Chọn ngày...");
                        }
                        else if (_element.DataType == "datetime")
                        {
                            this._tvValue.Text = CmmFunction.GetTitle("TEXT_CHOOSE_DATETIME", "Chọn ngày giờ...");
                        }
                        this._tvValue.SetTypeface(this._tvValue.Typeface, TypefaceStyle.Italic);
                        _tvValue.SetTextSize(ComplexUnitType.Sp, 12);
                    }
                    else
                    {
                        this._tvValue.Text = "";
                    }
                }
            }
            get { return this._tvValue.Text; }
        }

        public ControlDate(Activity _mainAct, LinearLayout _parentView, ViewElement _element) : base(_mainAct)
        {
            this._mainAct = _mainAct;
            this._parentView = _parentView;
            this._element = _element;
            InitializeComponent();
        }
        public override void InitializeComponent()
        {
            base.InitializeComponent();
            if (_lnContent != null)
            {
                _lnContent.Click += HandleTouchDown;
            }
        }
        public override void InitializeFrameView(LinearLayout frame)
        {
            if (_element.Hidden == true) // Check xem có ẩn view hay không
                return;

            base.InitializeFrameView(frame);
        }
        /// <summary>
        /// Event khi click vào Control
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="e"></param>
        protected virtual void HandleTouchDown(object sender, EventArgs e)
        {
            if (_parentView != null /* && _element.Enable == true*/)
            {
                MinionActionCore.OnElementFormClick(null, new MinionActionCore.ElementFormClick(_element));
            }
        }
        public override void SetProprety()
        {
            if (_element.ListProprety != null)
            {
                foreach (var item in _element.ListProprety)
                {
                    CmmDroidFunction.SetPropertyValueByNameCustom(_tvValue, item.Key, item.Value);
                }
            }
        }
        public override void SetEnable()
        {
            base.SetEnable();

            if (_element.Enable)
            {
                _tvValue.SetTextColor(new Color(ContextCompat.GetColor(_mainAct, Resource.Color.clBlueEnable)));
            }
        }
        public override void SetTitle()
        {
            base.SetTitle();

            _tvTitle.Text = _element.Title;

            if (_element.IsRequire && _element.Enable)
            {
                _tvTitle.Text += " (*)";
                CmmDroidFunction.SetTextViewHighlightControl(_mainAct, _tvTitle);
            }
        }
        public override void SetValue()
        {
            base.SetValue();

            Value = _element.Value;
        }
    }
}

File: VNABPMOP/app/src/main/java/com/vuthao/bpmop/core/component/ControlDate.java
package com.vuthao.bpmop.core.component;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.content.Intent;
import android.graphics.Typeface;
import android.util.TypedValue;
import android.view.View;
import android.widget.LinearLayout;

import androidx.core.content.ContextCompat;
import androidx.localbroadcastmanager.content.LocalBroadcastManager;

import com.google.gson.Gson;
import com.vuthao.bpmop.R;
import com.vuthao.bpmop.base.BroadcastUtility;
import com.vuthao.bpmop.base.Functions;
import com.vuthao.bpmop.base.MultipleClickGuard;
import com.vuthao.bpmop.base.model.dynamic.ViewElement;
import com.vuthao.bpmop.base.vars.VarsReceiver;
import com.vuthao.bpmop.detail.custom.DetailFunc;

public class ControlDate extends ControlBase {
    private ViewElement element;

    public ControlDate(Activity mainAct) {
        super(mainAct);
    }

    public ControlDate(Activity mainAct, LinearLayout parentView, ViewElement element) {
        super(mainAct);
        this.mainAct = mainAct;
        this.element = element;
        initializeComponent();
    }

    @Override
    public void setValue(String value) {
        if (!Functions.isNullOrEmpty(value)) {
            String customValue = "";

            customValue = DetailFunc.share.formatDateTimeControl(value, element.getDataType(), "yyyy-MM-dd HH:mm:ss");
            tvValue.setText(customValue);
        } else {
            if (element.isEnable()) {
                if (element.getDataType().equals("date")) {
                    tvValue.setText(Functions.share.getTitle("TEXT_CHOOSE_DATE", "Chọn ngày..."));
                } else if (element.getDataType().equals("datetime")) {
                    tvValue.setText(Functions.share.getTitle("TEXT_CHOOSE_DATETIME", "Chọn ngày giờ..."));
                }

                tvValue.setTypeface(tvValue.getTypeface(), Typeface.ITALIC);
                tvValue.setTextSize(TypedValue.COMPLEX_UNIT_SP, 12);
            }
        }
    }

    @Override
    public void initializeComponent() {
        super.initializeComponent();
        if (lnContent != null) {
            lnContent.setOnClickListener(new MultipleClickGuard(view -> handleClick(), 2000));
        }
    }

    public void handleClick() {
        Intent intent = new Intent();
        intent.setAction(VarsReceiver.FORMCLICK);
        intent.putExtra("element", new Gson().toJson(element));
        BroadcastUtility.send(mainAct, intent);
    }

    @Override
    public void initializeFrameView(LinearLayout frame) {
        if (element.isHidden()) return;
        super.initializeFrameView(frame);
    }

    @Override
    public void setProprety() {
    }

    @Override
    public void setEnable() {
        super.setEnable();

        if (element.isEnable()) {
            tvValue.setTextColor(ContextCompat.getColor(mainAct, R.color.clBlueEnable));
        }
    }

    @SuppressLint("SetTextI18n")
    @Override
    public void setTitle() {
        super.setTitle();

        tvTitle.setText(element.getTitle());
        if (element.isRequire() && element.isEnable()) {
            tvTitle.setText(tvTitle.getText() + " (*)");
            Functions.share.setTVHighlightControl(mainAct, tvTitle);
        }
    }

    @Override
    public void setValue() {
        super.setValue();
        setValue(element.getValue());
    }
}
